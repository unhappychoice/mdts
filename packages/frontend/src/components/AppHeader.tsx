import GitHubIcon from '@mui/icons-material/GitHub';
import SettingsIcon from '@mui/icons-material/Settings';
import { AppBar, Box, IconButton, Link, Toolbar, Tooltip } from '@mui/material';
import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';

interface AppHeaderProps {
  handleFileSelect: (path: string) => void;
  onSettingsClick: () => void;
}

const AppHeader: React.FC<AppHeaderProps> = ({
  handleFileSelect,
  onSettingsClick,
}) => {
  const dispatch = useDispatch();

  const handleFileSelectClick = useCallback(() => {
    handleFileSelect('');
  }, [dispatch, handleFileSelect]);

  return (
    <AppBar
      position="static"
      elevation={0}
      sx={(theme) => ({
        bgcolor: theme.palette.background.paper,
        color: theme.palette.text.primary,
        borderBottom: `1px solid ${theme.palette.divider}`,
      })}
    >
      <Toolbar>
        <Box sx={{ flexGrow: 1 }}>
          <Tooltip title="Top page">
            <IconButton disableRipple onClick={handleFileSelectClick} color="inherit">
              <img
                src="/logo.svg"
                alt="mdts logo"
                style={{ height: '56px', marginTop: '-8px', marginLeft: '-36px' }}
              />
            </IconButton>
          </Tooltip>
        </Box>
        <Tooltip title="Settings">
          <IconButton sx={{ mr: 2 }} onClick={onSettingsClick} color="inherit">
            <SettingsIcon />
          </IconButton>
        </Tooltip>
        <Tooltip title="GitHub Repository">
          <IconButton
            color="inherit"
            href="https://github.com/unhappychoice/mdts"
            target="_blank"
            rel="noopener"
            sx={{ mr: 2 }}
          >
            <GitHubIcon />
          </IconButton>
        </Tooltip>
        <Box sx={{ fontSize: '0.9rem', pt: '2px', fontFamily: 'monospace' }}>
          <Tooltip title="Changelog">
            <Link href="https://github.com/unhappychoice/mdts/blob/main/CHANGELOG.md" target="_blank" rel="noopener">
              v{process.env.APP_VERSION}
            </Link>
          </Tooltip>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default AppHeader;
