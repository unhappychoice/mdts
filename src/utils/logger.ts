type LogTag = 'CLI' | 'Server' | 'Livereload' | 'Announcement';

class Logger {
  private silent = false;

  setSilent(silent: boolean): void {
    this.silent = silent;
  }

  showLogo(): void {
    if (this.silent) {
      return;
    }

    /* eslint-disable @stylistic/max-len */
    console.log(`
[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m [39m[38;5;39m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m
[38;5;45m [39m[38;5;45m [39m[38;5;45m_[39m[38;5;45m [39m[38;5;45m_[39m[38;5;45m_[39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m_[39m[38;5;39m_[39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m_[39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m_[39m[38;5;33m [39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m [39m
[38;5;45m [39m[38;5;45m|[39m[38;5;45m [39m[38;5;45m'[39m[38;5;45m_[39m[38;5;45m [39m[38;5;39m\`[39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m [39m[38;5;39m\\[39m[38;5;39m [39m[38;5;39m/[39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m\`[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m_[39m[38;5;39m_[39m[38;5;33m/[39m[38;5;33m [39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m|[39m
[38;5;45m [39m[38;5;45m|[39m[38;5;45m [39m[38;5;45m|[39m[38;5;45m [39m[38;5;45m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m([39m[38;5;39m_[39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m_[39m[38;5;33m\\[39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m [39m[38;5;33m\\[39m
[38;5;45m [39m[38;5;45m|[39m[38;5;45m_[39m[38;5;45m|[39m[38;5;45m [39m[38;5;45m|[39m[38;5;39m_[39m[38;5;39m|[39m[38;5;39m [39m[38;5;39m|[39m[38;5;39m_[39m[38;5;39m|[39m[38;5;39m\\[39m[38;5;39m_[39m[38;5;39m_[39m[38;5;39m,[39m[38;5;39m_[39m[38;5;39m|[39m[38;5;39m\\[39m[38;5;39m_[39m[38;5;39m_[39m[38;5;33m|[39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m_[39m[38;5;33m/[39m
[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;45m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;39m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m[38;5;33m [39m
    `);
    /* eslint-enable @stylistic/max-len */
  }

  log(tag: LogTag, message: string, ...args: unknown[]): void {
    if (this.silent) {
      return;
    }
    this.getTagColor(tag)
      .then(chalk => {
        const paddedTag = tag.padEnd(12, ' '); // Pad to 12 characters
        const tagColored = chalk(` ${paddedTag} `);

        const formattedMessage = message.replace(/(https?:\/\/\S+)/g, (url) => {
          // Apply OSC 8 for clickable link and chalk for styling
          return `\x1b]8;;${url}\x1b\\${url}\x1b]8;;\x1b\\`;
        });

        console.log(`${tagColored} ${formattedMessage}`, ...args);
      });
  }

  error(message: string, ...args: unknown[]): void {
    import('chalk')
      .then((module) => module.default)
      .then(chalk => { console.error(chalk.bgRed.black(' Error'.padEnd(14, ' ')), message, ...args); });
  }

  private getTagColor(tag: LogTag) {
    return import('chalk').then((module) => module.default).then(chalk => {
      switch (tag) {
        case 'CLI':
          return chalk.bgGreen.black;
        case 'Server':
          return chalk.bgCyan.black;
        case 'Livereload':
          return chalk.bgBlue.black;
        case 'Announcement':
          return chalk.bgYellow.black;
        default:
          return chalk.bgWhite.black;
      }
    });
  }
}

export const logger = new Logger();
