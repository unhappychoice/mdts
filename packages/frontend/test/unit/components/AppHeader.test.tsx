import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import AppHeader from '../../../src/components/AppHeader';
import { createMockStore } from '../../utils';

describe('AppHeader', () => {
  let store;

  beforeEach(() => {
    store = createMockStore();
    store.dispatch = jest.fn();
  });

  test('renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AppHeader handleFileSelect={jest.fn()} onSettingsClick={jest.fn()} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('calls handleFileSelect when logo is clicked', () => {
    const handleFileSelectMock = jest.fn();
    render(
      <Provider store={store}>
        <AppHeader handleFileSelect={handleFileSelectMock} onSettingsClick={jest.fn()} />
      </Provider>
    );
    fireEvent.click(screen.getByText('mdts'));
    expect(handleFileSelectMock).toHaveBeenCalledWith('');
  });
});
