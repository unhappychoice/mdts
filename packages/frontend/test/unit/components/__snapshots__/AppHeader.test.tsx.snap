// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AppHeader renders correctly 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-83mobe-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
    >
      <div
        class="MuiBox-root css-1xvvp3o"
      >
        <button
          aria-label="Top page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1vqxir3-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <div
            style="height: 56px; margin-top: -8px; margin-left: -16px;"
          >
            <svg
              style="width: auto; height: 100%;"
              viewBox="0 0 180 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  id="grad"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="1"
                >
                  <stop
                    offset="0%"
                    stop-color="#1976d2"
                  />
                  <stop
                    offset="100%"
                    stop-color="#1976d2"
                  />
                </lineargradient>
                <style>
                  
            .text {
              font-family: 'Segoe UI', 'Arial', sans-serif;
              font-size: 42px;
              font-weight: 700;
              fill: url(#grad);
              letter-spacing: 2px;
            }
            .icon {
              fill: url(#grad);
            }
          
                </style>
              </defs>
              <g
                class="icon"
                transform="translate(10, 20)"
              >
                <path
                  d="M5 25 L15 45"
                  stroke="url(#grad)"
                  stroke-width="2"
                />
                <path
                  d="M25 25 L15 45"
                  stroke="url(#grad)"
                  stroke-width="2"
                />
                <rect
                  height="10"
                  rx="2"
                  width="10"
                  x="0"
                  y="20"
                />
                <rect
                  height="10"
                  rx="2"
                  width="10"
                  x="20"
                  y="20"
                />
                <rect
                  height="10"
                  rx="2"
                  width="10"
                  x="10"
                  y="40"
                />
              </g>
              <text
                class="text"
                x="60"
                y="70"
              >
                mdts
              </text>
            </svg>
          </div>
        </button>
      </div>
      <button
        aria-label="Settings"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-v8k5r1-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="SettingsIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
          />
        </svg>
      </button>
      <a
        aria-label="GitHub Repository"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-v8k5r1-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        href="https://github.com/unhappychoice/mdts"
        rel="noopener"
        tabindex="0"
        target="_blank"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="GitHubIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
          />
        </svg>
      </a>
      <div
        class="MuiBox-root css-gcae57"
      >
        <a
          aria-label="Changelog"
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
          data-mui-internal-clone-element="true"
          href="https://github.com/unhappychoice/mdts/blob/main/CHANGELOG.md"
          rel="noopener"
          target="_blank"
        >
          v
        </a>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
